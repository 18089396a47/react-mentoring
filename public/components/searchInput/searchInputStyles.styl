.search-input
  position: relative;
  margin-bottom: 15px;

  &__field
    border: 0;
    border-bottom: 1px solid #565656;
    height: 55px;
    line-height: 54px;
    font-size: 18px;
    padding: 0 25px;
    background: #000;
    color: #ccc;
    outline: 0;
    width: 100%;
    transition: border-bottom 200ms;

    &:focus,
    &:not(:placeholder-shown)
      border-bottom: 1px solid #f45162;

  &__icon
    position: absolute;
    width: 30px;
    height: 30px;
    top: 50%;
    right: 34px;
    margin-top: -13px;
    cursor: pointer;
    padding: 5px;
    background: url(./searchInputIcon.png) no-repeat;
    background-position: 6px;
