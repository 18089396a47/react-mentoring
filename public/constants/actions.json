{
  "searchStart": "SEARCH_START",
  "searchEnd": "SEARCH_END",
  "changeSearchInput": "CHANGE_SEARCH_INPUT",
  "updateSearchResults": "UPDATE_SEARCH_RESULTS"
}
