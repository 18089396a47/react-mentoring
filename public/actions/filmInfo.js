import { call, put, all, takeLatest } from 'redux-saga/effects';
import * as types from '../constants/actions';
import theMovieDb from 'themoviedb-javascript-library';
import * as commons from '../constants/commons';
import promiseWrapper from '../helpers/promiseWrapper';

export const searchFilmEnd = () => ({
    type: types.searchFilmEnd
});

export const searchFilmStart = (id, searchType) => ({
    type: types.searchFilmStart,
    id,
    searchType
});

function* fetchFilm({ id, searchType }) {
    const searchMethod = searchType === commons.searchTypeMovie ?
        theMovieDb.movies.getById : theMovieDb.tv.getById;

    try {
        const response = yield call(() => promiseWrapper(searchMethod, {
            id
        }));

        yield put(searchFilmEnd());
        yield put(updateFilmInfo(JSON.parse(response)))
    } catch (e) {
        yield put(searchFilmEnd());
    }
}

function* watchFetchFilm() {
    yield takeLatest(types.searchFilmStart, fetchFilm);
}

export function* filmInfoSaga() {
    yield all([
        watchFetchFilm()
    ]);
}

export const updateFilmInfo = (data) => ({
    type: types.updateFilmInfo,
    data
});
