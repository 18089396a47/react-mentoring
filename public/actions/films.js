import * as types from '../constants/actions';

export const updateSearchResults = (data) => ({
    type: types.updateSearchResults,
    data: JSON.parse(data)
});
