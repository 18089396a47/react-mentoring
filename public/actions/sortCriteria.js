import * as types from '../constants/actions';

export const changeSortCriteria = (sortCriteria) => ({
    type: types.changeSortCriteria,
    sortCriteria
});
