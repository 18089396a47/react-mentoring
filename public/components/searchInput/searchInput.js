import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import theMovieDb from 'themoviedb-javascript-library';
import './searchInputStyles';
import { search } from '../../actions/';

const SearchInput = ({ dispatch }) => {
    let input;

    function searchFilm(event) {
        event.preventDefault();
        const value = input.value.trim();
        
        if (value) {
            dispatch(search(value));
            input.value = '';
        }
    }

    return (
        <form className="search-input" onSubmit={searchFilm}>
            <input
                ref={node => { input = node; }}
                className="search-input__field"
                placeholder="Enter something..."
            />
            <div className="search-input__icon" onClick={searchFilm}></div>
        </form>
    );
};

SearchInput.propTypes = {
    dispatch: PropTypes.func.isRequired
};

export default connect()(SearchInput);
