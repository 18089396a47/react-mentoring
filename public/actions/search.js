import { call, put, all, takeLatest } from 'redux-saga/effects';
import * as types from '../constants/actions';
import theMovieDb from 'themoviedb-javascript-library';
import { updateSearchResults } from './films';
import * as commons from '../constants/commons';
import promiseWrapper from '../helpers/promiseWrapper';

export const changeSearchInput = (inputValue) => ({
    type: types.changeSearchInput,
    inputValue
});

export const searchQueryEnd = () => ({
    type: types.searchQueryEnd
});

export const searchQueryStart = (inputValue, searchType) => ({
    type: types.searchQueryStart,
    inputValue,
    searchType
});

function* fetchQuery({ inputValue, searchType }) {
    const searchMethod = searchType === commons.searchTypeMovie ?
        theMovieDb.search.getMovie : theMovieDb.search.getTv;

    try {
        const response = yield call(() => promiseWrapper(searchMethod, {
            query: inputValue
        }));

        yield put(searchQueryEnd());
        yield put(updateSearchResults(JSON.parse(response)))
    } catch (e) {
        yield put(searchQueryEnd());
    }
};

function* watchFetchQuery() {
    yield takeLatest(types.searchQueryStart, fetchQuery);
}

export const searchSimilarFilmEnd = () => ({
    type: types.searchSimilarFilmEnd
});

export const searchSimilarFilmStart = (id) => ({
    type: types.searchSimilarFilmStart,
    id
});

function* fetchSimilarFilms({ id }) {
    try {
        const response = yield call(() => promiseWrapper(theMovieDb.movies.getSimilarMovies, {
            id
        }));

        yield put(searchSimilarFilmEnd());
        yield put(updateSearchResults(JSON.parse(response)))
    } catch (e) {
        yield put(searchSimilarFilmEnd());
    }
};

function* watchFetchSimilarFilms() {
    yield takeLatest(types.searchSimilarFilmStart, fetchSimilarFilms);
}

export function* searchSaga() {
    yield all([
        watchFetchSimilarFilms(),
        watchFetchQuery()
    ]);
}

export const changeSearchType = (searchType) => ({
    type: types.changeSearchType,
    searchType
});
