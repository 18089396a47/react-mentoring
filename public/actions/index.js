import { all } from 'redux-saga/effects';

import { filmInfoSaga } from './filmInfo';
import { searchSaga } from './search';

export default function* rootSaga() {
    yield all([
        filmInfoSaga(),
        searchSaga()
    ]);
}

export * from './search';
export * from './sortCriteria';
export * from './filmInfo';
export * from './films';
