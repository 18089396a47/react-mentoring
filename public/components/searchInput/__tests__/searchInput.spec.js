import React from 'react';
import renderer from 'react-test-renderer';
import { SearchInput } from '../searchInput';
import { searchQueryStart, changeSearchInput } from '../../../actions/';

const mockDigest = jest.fn();

jest.mock('../../../actions/index', () => ({
    searchQueryStart: jest.fn(),
    changeSearchInput: jest.fn()
}));

let props;

describe('SearchInput component', () => {
    beforeEach(() => {
        props = {
            dispatch: mockDigest,
            searchType: 'searchType',
            inputValue: 'inputValue'
        };
    });

    it('SearchInput should be shown correctly', () => {
        const tree = renderer.create(
            <SearchInput {...props} />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('SearchInput should call \'changeSearchInput\' redux action on input keyDown', () => {
        const tree = shallow(<SearchInput {...props} />);
        const input = tree.find('.search-input__field');

        input.simulate('change', {
            target: {
                value: 'newInputValue'
            }
        });
        expect(mockDigest).toHaveBeenCalled();
        expect(changeSearchInput).toHaveBeenCalledWith('newInputValue');
    });

    it('SearchInput should call \'searchQueryStart\' redux action, history push on input keyDown', () => {
        props.history = {
            push: jest.fn()
        };
        const tree = shallow(<SearchInput {...props} />);

        tree.simulate('submit', { preventDefault: jest.fn() });
        expect(mockDigest).toHaveBeenCalled();
        expect(props.history.push).toHaveBeenCalledWith('/search/inputValue');
        expect(searchQueryStart).toHaveBeenCalledWith('inputValue', 'searchType');
    });
});
