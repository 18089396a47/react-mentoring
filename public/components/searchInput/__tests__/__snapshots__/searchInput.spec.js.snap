// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInput component SearchInput should be shown correctly 1`] = `
<form
  className="search-input"
  onSubmit={[Function]}
>
  <input
    className="search-input__field"
    onChange={[Function]}
    placeholder="Enter something..."
    value="inputValue"
  />
  <div
    className="search-input__icon"
    onClick={[Function]}
  />
</form>
`;
